In [139]:
%load_ext ipydex.displaytools
import sympy as sp

from sympy import I as j
from sympy import latex
from sympy import gcdex

The ipydex.displaytools extension is already loaded. To reload it, use:
  %reload_ext ipydex.displaytools


# Problem 1

## Part A

In [140]:
s, k, Q = sp.symbols('s k Q')

In [141]:
P = (4 * (s - 2)) / (s**2 - 2*s + 2) ##:

P := (4*s - 8)/(s**2 - 2*s + 2)

---


Transform $P(s)$ to $\bar{P}(k)$ under the mapping $s = \frac{1-k}{k}$ 

In [142]:
P_bar = P.subs({s: (1-k)/k}).factor().cancel() ##:

P_bar := -(12*k**2 - 4*k)/(5*k**2 - 4*k + 1)

---


In [143]:
# thus
n = P_bar.as_numer_denom()[0] ##:
d = P_bar.as_numer_denom()[1] ##:

n := -12*k**2 + 4*k

---


d := 5*k**2 - 4*k + 1

---


Using Euclid’s algorithm, find polynomials $x(k)$, $y(k)$ such that $nx + dy = 1$

In [144]:
x, y, _ = gcdex(n.as_poly(), d.as_poly())

In [145]:
x = x ##:
y = y ##:

x := Poly(35/8*k - 13/8, k, domain='QQ')

---


y := Poly(21/2*k + 1, k, domain='QQ')

---


In [146]:
x = (35/8)*k - 13/8 ##:
y = (21/2)*k + 1 ##:

x := 4.375*k - 1.625

---


y := 10.5*k + 1

---


In [147]:
N = n.subs({k: 1/(s+1)}).simplify() ##:
D = d.subs({k: 1/(s+1)}).simplify() ##:

X = x.subs({k: 1/(s+1)}).simplify() ##:
Y = y.subs({k: 1/(s+1)}).simplify() ##:

N := 4*(s - 2)/(s**2 + 2*s + 1)

---


D := (s**2 - 2*s + 2)/(s**2 + 2*s + 1)

---


X := (2.75 - 1.625*s)/(s + 1)

---


Y := (s + 11.5)/(s + 1)

---


In [148]:
# by the Youla-Kucera parametrization theorem:
_C = (X + D*Q) / (Y - N*Q) ##:
C = _C.cancel().simplify().collect(Q) ##:

_C := (Q*(s**2 - 2*s + 2)/(s**2 + 2*s + 1) + (2.75 - 1.625*s)/(s + 1))/(-4*Q*(s - 2)/(s**2 + 2*s + 1) + (s + 11.5)/(s + 1))

---


C := (Q*(1.0*s**2 - 2.0*s + 2.0) - 1.625*s**2 + 1.125*s + 2.75)/(Q*(8.0 - 4.0*s) + 1.0*s**2 + 12.5*s + 11.5)

---


## Part B

The steady state performance conditions state that:
* for $R(s) = \frac{1}{s}$, $e_{ss} = 0$ &rarr; $C$ has a pole at $s=0$
* for $r(t) = sin(3t), R(s) = \frac{3}{s^2 + 9}$, $e_{ss} = 0$ &rarr; $C$ has poles at $s=\pm 3j$

Let us define the denominator of $C$: $Y - NQ$ as $\Psi(s)$, then we have the three conditions:
* $\Psi(0) = 0$
* $\Psi(3j) = 0$
* $\Psi(-3j) = 0$

And Q must be in the form
$$
Q(s) = \frac{q_2s^2 + q_1s + q_0}{(s+3)^2}
$$

we solve these three equations with three unknowns $q_2, q_1, q_0$ as follows.

$$
\begin{gathered}
\Psi(s) =& - \frac{4 \left(s - 2\right) \left(q_{0} + q_{1} s + q_{2} s^{2}\right)}{\left(s + 3\right)^{2} \left(s^{2} + 2 s + 1\right)} + \frac{s + 11.5}{s + 1} \\
\Psi(0) =& \frac{8 q_{0}}{9} + 11.5 = 0 \\
\Psi(3j) =& - \frac{\left(-8 - 6 i\right) \left(-2 + 3 i\right) \left(q_{0} + 3 i q_{1} - 9 q_{2}\right)}{25 \left(3 + 3 i\right)^{2}} + \frac{\left(1 - 3 i\right) \left(11.5 + 3 i\right)}{10} = 0\\
\Psi(-3j) =& - \frac{\left(-8 + 6 i\right) \left(-2 - 3 i\right) \left(q_{0} - 3 i q_{1} - 9 q_{2}\right)}{25 \left(3 - 3 i\right)^{2}} + \frac{\left(1 + 3 i\right) \left(11.5 - 3 i\right)}{10} = 0
\end{gathered}
$$

The solutions are
$$
q_0: -12.9375, \quad q_1 = 12.4038, \quad q_2 = -4.61058
$$

Thus,
$$
Q(s) = \frac{- 4.61058 s^{2} + 12.4038 s - 12.9375}{\left(s + 3\right)^{2}}
$$

Plugging this $Q(s)$ to the equation for the controller $C$, we finally arrive at:
$$
C(s) = - \frac{6.236 s^{6} + 24.41 s^{5} + 30.21 s^{4} + 118.2 s^{3} + 6.101 s^{2} - 689 s + 10.13}{s^{6} + 42.94 s^{5} + 239.7 s^{4} + 719 s^{3} + 2076 s^{2} + 2992 s}
$$

In [149]:
q0, q1, q2 = sp.symbols('q0 q1 q2')

In [150]:
Q_ = (q2*s**2 + q1*s + q0) / ((s+3)**2) ##:

Q_ := (q0 + q1*s + q2*s**2)/(s + 3)**2

---


In [151]:
Psi = Y - N*Q_ ##:

Psi := -4*(s - 2)*(q0 + q1*s + q2*s**2)/((s + 3)**2*(s**2 + 2*s + 1)) + (s + 11.5)/(s + 1)

---


In [152]:
eq1 = Psi.subs({s: 0}) ##:
eq2 = Psi.subs({s: 3*j}) ##:
eq3 = Psi.subs({s: -3*j}) ##:

eq1 := 8*q0/9 + 11.5

---


eq2 := -(-8 - 6*I)*(-2 + 3*I)*(q0 + 3*I*q1 - 9*q2)/(25*(3 + 3*I)**2) + (1 - 3*I)*(11.5 + 3*I)/10

---


eq3 := -(-8 + 6*I)*(-2 - 3*I)*(q0 - 3*I*q1 - 9*q2)/(25*(3 - 3*I)**2) + (1 + 3*I)*(11.5 - 3*I)/10

---


In [153]:
sol = sp.solve([eq1, eq2, eq3], [q0, q1, q2]) ##:

sol := {q0: -12.9375000000000, q1: 12.4038461538462, q2: -4.61057692307692}

---


In [154]:
Q_sol = Q_.subs({q0: -12.9375, q1: 12.4038, q2: -4.61058}) ##:

Q_sol := (-4.61058*s**2 + 12.4038*s - 12.9375)/(s + 3)**2

---


In [155]:
C

(Q*(1.0*s**2 - 2.0*s + 2.0) - 1.625*s**2 + 1.125*s + 2.75)/(Q*(8.0 - 4.0*s) + 1.0*s**2 + 12.5*s + 11.5)

In [156]:
C_ = sp.N(C.subs(Q, Q_sol).cancel().collect(s), 4)
C_sol = C_.as_numer_denom()[0] / C_.as_numer_denom()[1] ##:

C_sol := (-6.236*s**6 - 24.41*s**5 - 30.21*s**4 - 118.2*s**3 - 6.101*s**2 + 689.0*s - 10.13)/(1.0*s**6 + 42.94*s**5 + 239.7*s**4 + 719.0*s**3 + 2076.0*s**2 + 2992.0*s)

---


In [157]:
print(latex(C_sol))

\frac{- 6.236 s^{6} - 24.41 s^{5} - 30.21 s^{4} - 118.2 s^{3} - 6.101 s^{2} + 689.0 s - 10.13}{1.0 s^{6} + 42.94 s^{5} + 239.7 s^{4} + 719.0 s^{3} + 2076.0 s^{2} + 2992.0 s}


# Problem 2

## Part 1

From Problem 1 we have:
$$
\begin{gathered}
N(s) = \frac{4 \left(s - 2\right)}{s^{2} + 2 s + 1}, \quad D(s) = \frac{s^{2} - 2 s + 2}{s^{2} + 2 s + 1} \\
X(s) = \frac{2.75 - 1.625 s}{s + 1}, \quad Y(s) = \frac{s + 11.5}{s + 1}
\end{gathered}
$$

We know that the robust stability condition requires:
$$
|| W_m T||_{\infty} \leq 1
$$

where $W_m$ is the multiplicative uncertainity bound and $T$ is the complementary sensitivity function. It is given that $W_m = \delta (s+1)$. We also know that $T = N(X + DQ)$. Thus the robust stability condition becomes:

$$
|| \delta (s+1) N (X + DQ_c) || _{\infty} \leq 1
$$

Taking the constant $\delta$ outside, we arrive at:
$$
\delta _{max} = \frac{1}{\gamma_{opt}}
$$

where
$$
\gamma_{opt} = \inf_{Q \in H_\infty} || (s+1)N (X + DQ_c) || _{\infty}
$$

We then transform this expression to the form:
$$
\gamma_{opt} = \inf_{Q \in H_\infty} || W - MQ || _{\infty}
$$

using inner-outer factorization, after the necessary computations we arrive at:
$$
M(s) = \frac{s^{2} - 2 s + 2}{s^{2} + 2 s + 1}, \quad W(s) = \frac{4 \left(2.75 - 1.625 s\right) \left(s - 2\right)}{s^{2} + 2 s + 1}, \quad Q(s) = Q_c(s)\frac{-4(s-2)}{s+1}
$$


To solve this using the _Nevallina-Pick interpolation_, we must construct the vectors $a$ and $b$. The vector $a$ is the zeros of $M$, which are $(1 \pm j)$. Thus $a = [\alpha_1, \alpha_2] = [1 + j, 1- j]$. The `b` vector is the value of $W$ at these points, thus $b = [W(\alpha_1), W(\alpha_2)] = [2+j, 2-j]$. Plugging these values into the script `NevPickNew.m`, with

```
>> a = [1+j, 1-j]; b = [2+j, 2-j];
>> [gopt, Qopt] = NevPickNew(a,b)
```

yields the results:
$$
\gamma_{opt} = 3.4495, \quad Q_{opt}(s) = \frac{3.4495 (s - 0.4495)}{s + 0.4495}
$$

Thus,
$$
Qc_{opt} = Q_{opt}\frac{s+1}{-4(s-2)} = - \frac{\left(s + 1\right) \left(3.4495 s - 1.55055025\right)}{\left(s + 0.4495\right) \left(4 s - 8\right)}
$$

Substituting this value of $Qc_{opt}$ in the expression for the controller $C$ yields:
$$
C(s) = \frac{X + DQ_c}{Y - NQ_c} = \frac{1.82 s^{3} - 8.05 s^{2} + 13.3 s - 1.86}{s^{3} - 0.849 s^{2} + 50.9 s - 7.24}
$$

In [158]:
Q_c = sp.symbols('Q_c')

In [159]:
W = (s+1)*X*N.collect(s) ##:

W := 4*(2.75 - 1.625*s)*(s - 2)/(s**2 + 2*s + 1)

---


In [160]:
W.subs(s, 1-j).simplify()

2.0 - 1.0*I

In [161]:
term2 = (s+1)*N*(D*Q_c).factor() ##:

term2 := 4*Q_c*(s - 2)*(s**2 - 2*s + 2)/((s + 1)*(s**2 + 2*s + 1))

---


In [162]:
M = (s**2 - 2*s + 2) / (s**2 + 2*s + 1) ##:

M := (s**2 - 2*s + 2)/(s**2 + 2*s + 1)

---


In [119]:
Q_opt = (3.4495*(s - 0.4495)) / (s + 0.4495) ##:

Q_opt := (3.4495*s - 1.55055025)/(s + 0.4495)

---


In [122]:
Q_copt = Q_opt * ((s+1)/(-4*(s-2))).cancel() ##:

Q_copt := -(s + 1)*(3.4495*s - 1.55055025)/((s + 0.4495)*(4*s - 8))

---


In [163]:
Q = sp.symbols('Q')
Copt = sp.N(C.subs(Q, Q_copt).simplify(), 3) ##:

Copt := (-9.95*s**4 + 19.6*s**3 + 8.32*s**2 - 28.0*s - 6.79)/(4.0*s**4 + 57.6*s**3 - 55.1*s**2 - 138.0*s - 28.9)

---


In [164]:
print(latex(Copt))

\frac{- 9.95 s^{4} + 19.6 s^{3} + 8.32 s^{2} - 28.0 s - 6.79}{4.0 s^{4} + 57.6 s^{3} - 55.1 s^{2} - 138.0 s - 28.9}
